File: src/test/java/mission/step3/MenuGroupServiceTest.java
package mission.step3;

import kitchenpos.application.MenuGroupService;
import kitchenpos.domain.MenuGroup;
import kitchenpos.domain.MenuGroupRepository;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.verify;

@ExtendWith(MockitoExtension.class)
class MenuGroupServiceTest {

    @Mock
    private MenuGroupRepository menuGroupRepository;

    @InjectMocks
    private MenuGroupService menuGroupService;

    @Test
    @DisplayName("메뉴 그룹을 생성한다")
    void create() {
        // given
        MenuGroup request = new MenuGroup();
        request.setName("음료");

        MenuGroup expected = new MenuGroup();
        expected.setId(UUID.randomUUID());
        expected.setName("음료");

        given(menuGroupRepository.save(any(MenuGroup.class)))
                .willReturn(expected);

        // when
        MenuGroup actual = menuGroupService.create(request);

        // then
        assertThat(actual.getName()).isEqualTo("음료");
        verify(menuGroupRepository).save(any(MenuGroup.class));
    }

    @Test
    @DisplayName("메뉴 그룹의 이름이 null이면 예외가 발생한다")
    void createWithNullName() {
        // given
        MenuGroup request = new MenuGroup();
        request.setName(null);

        // when & then
        assertThatThrownBy(() -> menuGroupService.create(request))
                .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    @DisplayName("메뉴 그룹의 이름이 비어있으면 예외가 발생한다")
    void createWithEmptyName() {
        // given
        MenuGroup request = new MenuGroup();
        request.setName("");

        // when & then
        assertThatThrownBy(() -> menuGroupService.create(request))
                .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    @DisplayName("모든 메뉴 그룹을 조회한다")
    void findAll() {
        // given
        MenuGroup group1 = new MenuGroup();
        group1.setId(UUID.randomUUID());
        group1.setName("음료");

        MenuGroup group2 = new MenuGroup();
        group2.setId(UUID.randomUUID());
        group2.setName("메인");

        given(menuGroupRepository.findAll())
                .willReturn(List.of(group1, group2));

        // when
        List<MenuGroup> menuGroups = menuGroupService.findAll();

        // then
        assertThat(menuGroups).hasSize(2);
        assertThat(menuGroups.get(0).getName()).isEqualTo("음료");
        assertThat(menuGroups.get(1).getName()).isEqualTo("메인");
        verify(menuGroupRepository).findAll();
    }
}

File: src/test/kotlin/kitchenops/application/MenuGroupServiceTest.kt
package kitchenops.application

import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import kitchenpos.application.MenuGroupService
import kitchenpos.domain.MenuGroup
import kitchenpos.domain.MenuGroupRepository
import spec.BaseUnitSpec
import java.util.*

internal class MenuGroupServiceTest : BaseUnitSpec({

    val menuGroupRepository = mockk<MenuGroupRepository>()
    val sut = MenuGroupService(menuGroupRepository)

    context("메뉴 그룹을 생성할 수 있다") {
        test("이름을 지정하지 않으면 생성할 수 없다") {
            // given
            val request = MenuGroup().apply { name = null }

            // when
            val actual = runCatching { sut.create(request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("이름이 비어있으면 생성할 수 없다") {
            // given
            val request = MenuGroup().apply { name = "" }

            // when
            val actual = runCatching { sut.create(request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("생성한다") {
            // given
            val requestName = "myName"
            val request = MenuGroup().apply { name = requestName }

            every { menuGroupRepository.save(match { it.name == requestName }) } returns request

            // when
            val actual = sut.create(request)

            // then
            actual.name shouldBe "myName"

            verify(exactly = 1) { menuGroupRepository.save(any()) }
        }
    }

    context("메뉴 그룹을 전체 조회할 수 있다") {
        test("전체 조회한다") {
            // given
            val menuGroup1 = MenuGroup().apply { id = UUID.randomUUID() }
            val menuGroup2 = MenuGroup().apply { id = UUID.randomUUID() }

            every { menuGroupRepository.findAll() } returns listOf(menuGroup1, menuGroup2)

            // when
            val actual = sut.findAll()

            // then
            actual shouldBe listOf(menuGroup1, menuGroup2)
        }
    }
})
